<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Chubing Zeng" />

<meta name="date" content="2019-05-17" />

<title>Getting started with xtune</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Getting started with <code>xtune</code></h1>
<h4 class="author"><em>Chubing Zeng</em></h4>
<h4 class="date"><em>2019-05-17</em></h4>



<div id="purpose-of-this-vignette" class="section level3">
<h3>Purpose of this vignette</h3>
<p>This vignette is a tutorial on how to use the <code>xtune</code> package to fit differential-shrinkage penalized regression models based on external information.</p>
<p>In this tutorial the following points are going to be viewed:</p>
<ul>
<li>The overall idea of <code>xtune</code> model</li>
<li>Three examples of external information Z</li>
<li>How to fit the model</li>
<li>Two special cases of Z</li>
</ul>
</div>
<div id="xtune-overview" class="section level3">
<h3><code>xtune</code> Overview</h3>
<p>The main usage of <code>xtune</code> is to tune multiple shrinkage parameters in Lasso and Ridge regression, based on external information.</p>
<p>The classical penalized regression uses a single penalty parameter <span class="math inline">\(\lambda\)</span> that applies equally to all regression coefficients to control the amount of regularization in the model. And the single penalty parameter tuning is typically performed using cross-validation.</p>
<p>Here we apply an individual shrinkage parameter <span class="math inline">\(\lambda_j\)</span> to each regression coefficient <span class="math inline">\(\beta_j\)</span>. And the vector of shrinkage parameters <span class="math inline">\(\lambda s = (\lambda_1,...,\lambda_p)\)</span> is guided by external information <span class="math inline">\(Z\)</span>. In specific, <span class="math inline">\(\lambda\)</span>s is modeled as a log-linear function of <span class="math inline">\(Z\)</span>. Better prediction accuracy for penalized regression models may be achieved by allowing individual shrinkage for each regression coefficients based on external information.</p>
<p>To tune the differential shrinkage parameter vector <span class="math inline">\(\lambda s = (\lambda_1,...,\lambda_p)\)</span>, we employ an Empirical Bayes approach by specifying LASSO and Ridge to their random-effect formulation. Once the tuning parameters <span class="math inline">\(\lambda\)</span>s are estimated, and therefore the penalties known, the regression coefficients are obtained using the <code>glmnet</code> package.</p>
<p>The response variable can be either quantitative or a factor with two levels. For a binary response variable, a sparse LDA procedure is used to get the fisher discriminant direction and the discriminant function for a new data point.</p>
<p>The sparse LDA procedure is described in this paper:</p>
<p>Mai, Qing &amp; Zou, Hui &amp; Yuan, Ming. (2012). A direct approach to sparse discriminant analysis in ultra-high dimensions. Biometrika. 99. 10.2307/41720670.</p>
<p>Utilities for carrying out post-fitting summarization and prediction are also provided.</p>
</div>
<div id="examples" class="section level3">
<h3>Examples</h3>
<p>This file uses three simulated examples to illustrate the usage and syntax of <code>xtune</code>. The first example gives users a general sense of the data structure and model fitting process. The second and third examples use simulated data in concrete scenarios to illustrate the usage of the package. In the second example <code>diet</code>, we provide simulated data to mimic the dietary example described in this paper:</p>
<p>S. Witte, John &amp; Greenland, Sander &amp; W. Haile, Robert &amp; L. Bird, Cristy. (1994). Hierarchical Regression Analysis Applied to a Study of Multiple Dietary Exposures and Breast Cancer. Epidemiology (Cambridge, Mass.). 5. 612-21. 10.1097/00001648-199411000-00009.</p>
<p>And in the third example <code>gene</code>, we provide simulated data to mimic the bone density data published in the European Bioinformatics Institute (EMBL-EBI) ArrayExpress repository, ID: E-MEXP-1618.</p>
<div id="general-example" class="section level4">
<h4>General Example</h4>
<p>In the first example, <span class="math inline">\(Y\)</span> is a <span class="math inline">\(n = 100\)</span>-dimensional continuous observed outcome vector, <span class="math inline">\(X\)</span> is matrix of <span class="math inline">\(p\)</span> potential predictors observed on the <span class="math inline">\(n\)</span> observations, and <span class="math inline">\(Z\)</span> is a set of <span class="math inline">\(q = 4\)</span> external features available for the <span class="math inline">\(p = 300\)</span> predictors.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(xtune)
<span class="kw">data</span>(<span class="st">&quot;example&quot;</span>)
X &lt;-<span class="st"> </span>example<span class="op">$</span>X; Y &lt;-<span class="st"> </span>example<span class="op">$</span>Y; Z &lt;-<span class="st"> </span>example<span class="op">$</span>Z
<span class="kw">dim</span>(X);<span class="kw">dim</span>(Z)
<span class="co">#&gt; [1] 100 300</span>
<span class="co">#&gt; [1] 300   4</span></code></pre></div>
<p>Each column of Z contains information about the predictors in design matrix X. The number of rows in Z equals to the number of predictors in X.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">X[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]
<span class="co">#&gt;               Predictor_1 Predictor_2 Predictor_3 Predictor_4 Predictor_5</span>
<span class="co">#&gt; Observation_1  -0.7667960   0.9212806   2.0149030  0.79004563  -1.4244699</span>
<span class="co">#&gt; Observation_2  -0.8164583  -0.3144157  -0.2253684  0.08712746  -1.0296026</span>
<span class="co">#&gt; Observation_3  -0.1415352   0.6623149  -1.0398456  1.87611212   0.7340254</span>
<span class="co">#&gt;               Predictor_6 Predictor_7 Predictor_8 Predictor_9 Predictor_10</span>
<span class="co">#&gt; Observation_1  -0.9529327  -0.9344928 -0.32964818   0.4486023  -0.70894600</span>
<span class="co">#&gt; Observation_2   2.2546851   0.2732793 -0.03852896   1.3830463   0.03070716</span>
<span class="co">#&gt; Observation_3   1.7534320   0.3263808  0.09564893  -2.2104531   0.22615224</span></code></pre></div>
<p>The external information is encoded as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Z[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,]
<span class="co">#&gt;              External_variable_1 External_variable_2 External_variable_3</span>
<span class="co">#&gt; Predictor_1                    1                   0                   0</span>
<span class="co">#&gt; Predictor_2                    1                   0                   0</span>
<span class="co">#&gt; Predictor_3                    0                   1                   0</span>
<span class="co">#&gt; Predictor_4                    0                   1                   0</span>
<span class="co">#&gt; Predictor_5                    0                   0                   1</span>
<span class="co">#&gt; Predictor_6                    0                   0                   1</span>
<span class="co">#&gt; Predictor_7                    0                   0                   0</span>
<span class="co">#&gt; Predictor_8                    0                   0                   0</span>
<span class="co">#&gt; Predictor_9                    0                   0                   0</span>
<span class="co">#&gt; Predictor_10                   0                   0                   0</span>
<span class="co">#&gt;              External_variable_4</span>
<span class="co">#&gt; Predictor_1                    0</span>
<span class="co">#&gt; Predictor_2                    0</span>
<span class="co">#&gt; Predictor_3                    0</span>
<span class="co">#&gt; Predictor_4                    0</span>
<span class="co">#&gt; Predictor_5                    0</span>
<span class="co">#&gt; Predictor_6                    0</span>
<span class="co">#&gt; Predictor_7                    1</span>
<span class="co">#&gt; Predictor_8                    1</span>
<span class="co">#&gt; Predictor_9                    1</span>
<span class="co">#&gt; Predictor_10                   1</span></code></pre></div>
<p>Here, each variable in Z is a binary variable. <span class="math inline">\(Z_{jk}\)</span> indicates if <span class="math inline">\(Predictor_j\)</span> has <span class="math inline">\(ExternalVariable_k\)</span> or not. This Z is an example of (non-overlapping) grouping of predictors. Predictor 1 and 2 belongs to group 1; predictor 3 and 4 belongs to group 2; predictor 5 and 6 belongs to group 3, and the rest of the predictors belongs to group 4.</p>
<p>To fit a differential-shrinkage lasso model to this data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.example1 &lt;-<span class="st"> </span><span class="kw">xtune</span>(X,Y,Z)
<span class="co">#&gt; Z provided, start estimating individual tuning parameters</span></code></pre></div>
<p>The individual penalty parameters are returned by</p>
<pre><code>fit.example1$penalty.vector</code></pre>
<p>In this example, predictors in each group get different estimated penalty parameters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">unique</span>(fit.example1<span class="op">$</span>penalty.vector)
<span class="co">#&gt;                   [,1]</span>
<span class="co">#&gt; Predictor_1 0.00768154</span>
<span class="co">#&gt; Predictor_3 0.01407634</span>
<span class="co">#&gt; Predictor_5 0.03259048</span>
<span class="co">#&gt; Predictor_7 0.87835961</span></code></pre></div>
<p>Coefficient estimates and predicted values and can be obtained via <code>predict</code> and <code>coef</code>:</p>
<pre><code>coef(fit.example1)
predict(fit.example1, newX = X)</code></pre>
<p>The <code>mse</code> function can be used to get the mean square error (MSE) between prediction values and true values.</p>
<pre><code>mse(predict(fit.example1, newX = X), Y)</code></pre>
</div>
<div id="dietary-example" class="section level4">
<h4>Dietary example</h4>
<p>Suppose we want to predict a person’s weight loss (binary outcome) using his/her weekly dietary intake. Our external information Z could incorporate information about the levels of relevant food constituents in the dietary items.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(diet)
<span class="kw">head</span>(diet<span class="op">$</span>DietItems)
<span class="co">#&gt;      Milk Margarine Eggs Apples Lettuce Celery Hot dogs Liver Dark bread</span>
<span class="co">#&gt; [1,]    1         1    4      1       2      1        1     0          0</span>
<span class="co">#&gt; [2,]    1         0    0      0       2      4        0     0          2</span>
<span class="co">#&gt; [3,]    0         1    2      3       1      3        0     0          4</span>
<span class="co">#&gt; [4,]    0         2    0      1       0      1        1     0          3</span>
<span class="co">#&gt; [5,]    0         1    1      3       1      1        0     2          2</span>
<span class="co">#&gt; [6,]    2         1    2      1       3      0        1     2          2</span>
<span class="co">#&gt;      Pasta Beer Liquor Cookies Bran</span>
<span class="co">#&gt; [1,]     1    0      2       3    4</span>
<span class="co">#&gt; [2,]     0    2      1       0    3</span>
<span class="co">#&gt; [3,]     0    1      2       1    1</span>
<span class="co">#&gt; [4,]     1    1      1       1    2</span>
<span class="co">#&gt; [5,]     2    0      0       0    3</span>
<span class="co">#&gt; [6,]     1    1      0       2    1</span>
<span class="kw">head</span>(diet<span class="op">$</span>weightloss)
<span class="co">#&gt; [1] 0 1 1 1 1 1</span></code></pre></div>
<p>The external information Z in this example is:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(diet<span class="op">$</span>NuitritionFact)
<span class="co">#&gt; NULL</span></code></pre></div>
<p>In this example, Z is not a grouping of the predictors. The idea is that the nutrition facts about the dietary items might give us some information on the importance of each predictor in the model.</p>
<p>Similar to the previous example, the xtune model could be fit by:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.diet =<span class="st"> </span><span class="kw">xtune</span>(<span class="dt">X =</span> diet<span class="op">$</span>DietItems,<span class="dt">Y=</span>diet<span class="op">$</span>weightloss,<span class="dt">Z =</span> diet<span class="op">$</span>NuitritionFact, <span class="dt">family=</span><span class="st">&quot;binary&quot;</span>)
<span class="co">#&gt; No Z matrix provided, only a single tuning parameter will be estimated using empirical Bayes tuning</span></code></pre></div>
<p>Each dietary predictor is estimated an individual tuning parameter based on their nutrition fact.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.diet<span class="op">$</span>penalty.vector
<span class="co">#&gt;  [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></code></pre></div>
<p>To make prediction using the trained model</p>
<pre><code>predict(fit.diet,newX = diet$DietItems)</code></pre>
<p>The above code returns the predicted probabilities (scores). To make a class prediction, use the <code>type = &quot;class&quot;</code> option and also provide the original X and Y used for the fit.</p>
<pre><code>pred_class &lt;- predict(fit.diet,newX = diet$DietItems,type = &quot;class&quot;,X = diet$DietItems,Y=diet$weightloss)</code></pre>
<p>The <code>misclassification()</code> function can be used to extract the misclassification rate. The prediction AUC can be calculated using the auc() function from the AUC package.</p>
<pre><code>misclassification(pred_class,true = diet$weightloss)</code></pre>
</div>
<div id="gene-expression-data-example" class="section level4">
<h4>Gene expression data example</h4>
<p>The <code>gene</code> data contains simulated gene expression data. The dimension of data is <span class="math inline">\(50\times 200\)</span>. The outcome Y is continuous (bone mineral density). The external information is four previous study results that identify the biological importance of genes. For example <span class="math inline">\(Z_{jk}\)</span> means whether <span class="math inline">\(gene_j\)</span> is identified to be biologically important in previous study <span class="math inline">\(k\)</span> result. <span class="math inline">\(Z_{jk} = 1\)</span> means that gene <span class="math inline">\(j\)</span> is identified by previous study <span class="math inline">\(k\)</span> result and <span class="math inline">\(Z_{jk} = 0\)</span> means that gene <span class="math inline">\(j\)</span> is not identified to be important by previous study <span class="math inline">\(k\)</span> result.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(gene)
gene<span class="op">$</span>GeneExpression[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]
<span class="co">#&gt;          Gene_1     Gene_2     Gene_3     Gene_4     Gene_5</span>
<span class="co">#&gt; [1,] -0.7667960  1.7520578  0.9212806 -0.6273008  2.0149030</span>
<span class="co">#&gt; [2,] -0.8164583 -0.5477714 -0.3144157 -0.8796116 -0.2253684</span>
<span class="co">#&gt; [3,] -0.1415352 -0.8585257  0.6623149 -0.3053110 -1.0398456</span>
gene<span class="op">$</span>PreviousStudy[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,]
<span class="co">#&gt;        Identified by previous study 1 Identified by previous study 2</span>
<span class="co">#&gt; Gene_1                              0                              0</span>
<span class="co">#&gt; Gene_2                              0                              0</span>
<span class="co">#&gt; Gene_3                              0                              0</span>
<span class="co">#&gt; Gene_4                              0                              0</span>
<span class="co">#&gt; Gene_5                              0                              0</span>
<span class="co">#&gt;        Identified by previous study 3 Identified by previous study 4</span>
<span class="co">#&gt; Gene_1                              0                              0</span>
<span class="co">#&gt; Gene_2                              0                              0</span>
<span class="co">#&gt; Gene_3                              0                              0</span>
<span class="co">#&gt; Gene_4                              0                              0</span>
<span class="co">#&gt; Gene_5                              0                              0</span></code></pre></div>
<p>A gene can be identified to be important by several previous study results, therefore the external information Z in this example can be seen as an overlapping group of variables.</p>
<p>Model fitting:</p>
<pre><code>fit.gene = xtune(X = gene$GeneExpression,Y=gene$bonedensity,Z = gene$PreviousStudy)</code></pre>
<p>The rest of the steps are the same as the previous two examples.</p>
</div>
</div>
<div id="two-special-cases" class="section level3">
<h3>Two special cases</h3>
<div id="no-external-information-z" class="section level4">
<h4>No external information Z</h4>
<p>If you just want to tune a single penalty parameter using empirical Bayes tuning, simply do not provide Z in the <code>xtune()</code> function. If no external information Z is provided, the function will perform empirical Bayes tuning to choose the single penalty parameter in penalized regression, as an alternative to cross-validation. For example</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.eb &lt;-<span class="st"> </span><span class="kw">xtune</span>(X,Y)
<span class="co">#&gt; No Z matrix provided, only a single tuning parameter will be estimated using empirical Bayes tuning</span></code></pre></div>
<p>The estimated tuning parameter is:</p>
<pre><code>fit.eb$lambda</code></pre>
</div>
<div id="z-as-an-identity-matrix" class="section level4">
<h4>Z as an identity matrix</h4>
<p>If you provide an identity matrix as external information Z to <code>xtune()</code>, the function will estimate a separate tuning parameter <span class="math inline">\(\lambda_j\)</span> for each regression coefficient <span class="math inline">\(\beta_j\)</span>. Note that this is not advised when the number of predictors <span class="math inline">\(p\)</span> is very large.</p>
<p>Using the dietary example, the following code would estimate a separate penalty parameter for each coefficient.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
Z_iden =<span class="st"> </span><span class="kw">diag</span>(<span class="kw">ncol</span>(diet<span class="op">$</span>DietItems))
fit.diet.identity =<span class="st"> </span><span class="kw">xtune</span>(diet<span class="op">$</span>DietItems,diet<span class="op">$</span>weightloss,Z_iden)
<span class="co">#&gt; Z provided, start estimating individual tuning parameters</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.diet.identity<span class="op">$</span>penalty.vector
<span class="co">#&gt;               [,1]</span>
<span class="co">#&gt;  [1,] 5.490105e-02</span>
<span class="co">#&gt;  [2,] 8.796161e-02</span>
<span class="co">#&gt;  [3,] 6.863956e+00</span>
<span class="co">#&gt;  [4,] 2.056148e-02</span>
<span class="co">#&gt;  [5,] 2.323703e+01</span>
<span class="co">#&gt;  [6,] 1.912083e-02</span>
<span class="co">#&gt;  [7,] 1.970638e-02</span>
<span class="co">#&gt;  [8,] 3.083099e-01</span>
<span class="co">#&gt;  [9,] 5.411627e-03</span>
<span class="co">#&gt; [10,] 3.176205e+00</span>
<span class="co">#&gt; [11,] 4.214480e-02</span>
<span class="co">#&gt; [12,] 1.117616e-01</span>
<span class="co">#&gt; [13,] 3.396484e+02</span>
<span class="co">#&gt; [14,] 5.695925e-02</span></code></pre></div>
<p>A predictor is excluded from the model (regression coefficient equals to zero) if its corresponding penalty parameter is estimated to be infinity.</p>
</div>
</div>
<div id="conclusion" class="section level3">
<h3>Conclusion</h3>
<p>We presented the main usage of <code>xtune</code> package. For more details about each function, please go check the package documentation. If you would like to give us feedback or report issue, please tell us on <a href="https://github.com/ChubingZeng/xtune">Github</a>.</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
